cmd/lakectl/cmd/annotate.go

```diff
@@ -0,0 +1,82 @@
+package cmd
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"github.com/treeverse/lakefs/pkg/api"
+)
+
+const annotateTemplate = `{{  $val := .Commit}}
+{{ $.Object|ljust 15}} {{ $val.Committer|ljust 20 }} {{ $val.Id | printf "%.16s"|ljust 20 }} {{ $val.CreationDate|date }}  {{ $.CommitMessage |ljust 30 }}
+`
+const messageSize = 200
+const ellipsis = "..."
+
+var annotateCmd = &cobra.Command{
+	Use:     "annotate <path uri>",
+	Short:   "List entries under a given path, annotating each with the latest modifying commit",
+	Aliases: []string{"blame"},
+	Args:    cobra.ExactArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		pathURI := MustParsePathURI("path", args[0])
+		client := getClient()
+		pfx := api.PaginationPrefix(*pathURI.Path)
+		context := cmd.Context()
+		res, err := client.ListObjectsWithResponse(context, pathURI.Repository, pathURI.Ref, &api.ListObjectsParams{Prefix: &pfx})
+		DieOnResponseError(res, err)
+		var from string
+		for {
+			params := &api.ListObjectsParams{
+				Prefix: &pfx,
+				After:  api.PaginationAfterPtr(from),
+			}
+			resp, err := client.ListObjectsWithResponse(context, pathURI.Repository, pathURI.Ref, params)
+			DieOnResponseError(resp, err)
+			for _, obj := range resp.JSON200.Results {
+				logCommitsParams := &api.LogCommitsParams{
+					Amount:  api.PaginationAmountPtr(1),
+					Objects: &[]string{obj.Path},
+				}
+				res, err := client.LogCommitsWithResponse(context, pathURI.Repository, pathURI.Ref, logCommitsParams)
+				DieOnResponseError(res, err)
+				if len(res.JSON200.Results) > 0 {
+					data := struct {
+						Commit        api.Commit
+						Object        string
+						CommitMessage string
+					}{
+						Commit:        res.JSON200.Results[0],
+						Object:        obj.Path,
+						CommitMessage: splitOnNewLine(stringTrimLen((res.JSON200.Results[0].Message), messageSize)),
+					}
+					Write(annotateTemplate, data)
+				} else {
+					fmt.Println("Error")
```

Comment:
1. We should not just print "Error" to the user - no commit information as part of the entry or something that they will understand we don't have commit log information for this record.
2. Skipping the record is also not an option as the entry is there and we don't want the user to think that an object is missing.

Reply:
Fixed